File: 0841. Keys and Rooms/0841_myans_DFS.cpp
class Solution
{
public:
    bool canVisitAllRooms(vector<vector<int>> &rooms)
    {
        int n = rooms.size();
        vector<bool> can_enter_room(n, false);
        stack<int> s;

        s.push(0);

        while (!s.empty())
        {
            int room_num = s.top();
            s.pop();

            if (!can_enter_room[room_num])
            {
                can_enter_room[room_num] = true;
                for (int &room : rooms[room_num])
                {
                    s.push(room);
                }
            }
        }

        for (int i = 0; i < n; i++)
        {
            if (!can_enter_room[i])
                return false;
        }

        return true;
    }
};

File: 0841. Keys and Rooms/0841_myans_DFS.go
func canVisitAllRooms(rooms [][]int) bool {
    n := len(rooms)
    canEnterRoom := make([]bool, n)
    stack := []int{0}

    for len(stack) > 0 {
        roomNum := stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        if !canEnterRoom[roomNum] {
            canEnterRoom[roomNum] = true
            for _, room := range rooms[roomNum] {
                stack = append(stack, room)
            }
        }
    }

    for i := 0; i < n; i++ {
        if !canEnterRoom[i] {
            return false
        }
    }

    return true
}